python/tvm/relax/dpl/rewrite.py

```diff
@@ -15,16 +15,154 @@
 # specific language governing permissions and limitations
 # under the License.
 """APIs for pattern-based rewriting."""
-from typing import Dict, Callable
+
+from typing import Dict, Callable, Union
+
+from tvm.ir import IRModule
+from tvm.runtime import Object
+from tvm._ffi import register_object
+
 from .pattern import DFPattern
 from .context import PatternContext
-
 from ..expr import Expr, Function, Var
 from . import _ffi as ffi
 
 
+@register_object("relax.dpl.ExprRewriter")
+class ExprRewriter(Object):
```

Comment:
Can we rename it to something like `ExprPatternMatchingRewriter`? `ExprRewriter` sounds too generic even though it is only about the pattern matching unless I'm missing something.

Reply:
Hmm.  I definitely agree on it being too generic of a name, but I think that specifying it as `PatternMatching` would be a bit too specific of a name.

For example, the existing `CombineParallelMatmul` pass couldn't be expressed as a pattern-matching rule (cannot express arbitrarily many matmuls with a shared RHS), but could be expressed as a subclass of `ExprRewriter`.  Doing so would then allow `CombineParallelMatmul` to be composed together with other `ExprRewriter` subclasses into a single optimization pass.

Reply:
Thinking on it a bit more, I think `PatternMatchingRewriter` would be a reasonable name.  Because the C++ API must return a `Map<Var,Expr>`, it is restricted to some sort of pattern-match.  That pattern may be represented in a specific `DFPattern` object, or it could be implicit in the function used to identify a match, but they're still restricted to pattern-matching.

Reply:
And names are all updated, running through CI.